### ISSUE TITLE
Reorder sparse matrices before iterative linear solve

### ISSUE THREAD

**Issue by dbeurle** [dbeurle/neon#135]
It might be beneficial to reorder the matrices before we pass to the CG / BiCGStab solvers, see https://www.sciencedirect.com/science/article/abs/pii/S0141933111000652

@shadialameddin Thoughts?

I think the node ordering out of the box is pretty bad.

**Comment by shadisharba** [comment]
Very interesting. It'll also be great when using these solvers with a modified Newton-Raphson scheme or latin :)
Did you find open source work related to this, I didn't?

**Comment by dbeurle** [comment]
Nah I have never seen any implementations of this in the wild.  It makes sense though.

**Comment by shadisharba** [comment]
[sparse_matrix.txt](https://github.com/dbeurle/neon/files/2732773/sparse_matrix.txt)

**Comment by shadisharba** [comment]
Just had an idea, as long as the matrix is assembled only once and then used in the iterative solver, why not restrict this matrix to a lower dimensional space where you can do the iterative scheme and expand the solution back to the original space?

Maybe this is already done inside the iterative solver, what do you think?

**Comment by dbeurle** [comment]
Can you elaborate a bit?  Are you suggesting using some kind of model order reduction?

**Comment by dbeurle** [comment]
Some notes for implementation:

## Row major (CSR) layout

- `innerIndexPtr()` returns the compressed row indices
- `outerIndexPtr()` returns the column indices

## Column major (CSC) layout

The opposite of above.

A sample of the results
```
Original sparse matrix A:
Nonzero entries:
(1,0) (1,4) (1,1) (1,2) (1,5) (1,7) (1,1) (1,2) (1,4) (1,3) (1,6) (1,0) (1,2) (1,4) (1,1) (1,5) (1,7) (1,3) (1,6) (1,1) (1,5) (1,7) 

Outer pointers:
0 2 6 9 11 14 17 19  $

1 0 0 0 1 0 0 0 
0 1 1 0 0 1 0 1 
0 1 1 0 1 0 0 0 
0 0 0 1 0 0 1 0 
1 0 1 0 1 0 0 0 
0 1 0 0 0 1 0 1 
0 0 0 1 0 0 1 0 
0 1 0 0 0 1 0 1 

Reordered matrix:
1 1 0 0 0 0 0 0 
1 1 0 0 0 0 0 0 
0 0 1 1 1 0 0 0 
0 0 1 1 1 0 0 0 
0 0 1 1 1 1 0 0 
0 0 0 0 1 1 1 0 
0 0 0 0 0 1 1 1 
0 0 0 0 0 0 1 1 


```
